#program path
fastQC = "/opt/nasapps/production/fastqc/0.11.2/bin/fastqc"
trimmomatic = "/is2/projects/CCR-SF/active/Software/tools/trimmomatic/Trimmomatic-0.30/trimmomatic-0.30.jar"
adapters = "/is2/projects/CCR-SF/active/RefGenomes/illumina_adapters/TruSeq_and_nextera_adapters.fa"
markedup = "/opt/nasapps/production/picard/1.84/MarkDuplicates.jar"
fastq_screen = "/opt/nasapps/development/fastq_screen/0.5.1/fastq_screen"
adapters = "/is2/projects/CCR-SF/active/RefGenomes/illumina_adapters/TruSeq_and_nextera_adapters.fa"
addreadgroup = "/opt/nasapps/production/picard/1.84/AddOrReplaceReadGroups.jar"
bbtools = "/is2/projects/CCR-SF/active/Software/tools/BBMap/bbmap/reformat.sh"
conf = "/is2/projects/CCR-SF/active/RefGenomes/univec/fastq_screen.conf"
fastq_screen = "/opt/nasapps/development/fastq_screen/0.5.1/fastq_screen"
fastQC = "/opt/nasapps/production/fastqc/0.11.2/bin/fastqc"
IlluQC = "/is2/projects/CCR-SF/active/Software/tools/NGSQCToolkit/2.3/QC/IlluQC_PRLL.pl"
kallisto = "/opt/nasapps/development/kallisto/0.42.4/bin/kallisto"
kraken = "/opt/nasapps/development/kraken/0.10.4/bin/kraken"
krakendb = "/opt/nasapps/development/kraken/0.10.4/minikraken_20141208"
krona = "/is2/projects/CCR-SF/active/Software/tools/Krona/v2.7/bin/ktImportTaxonomy"
markedup = "/opt/nasapps/production/picard/1.84/MarkDuplicates.jar"
multiqc = "/opt/nasapps/development/multiqc/0.7dev/bin/multiqc"
NGSQC_ADAPTER = "/is2/projects/CCR-SF/active/RefGenomes/illumina_adapters/TruSeq_and_nextera_adapters.ngsqc.temp"
ontarget = "/is2/projects/CCR-SF/active/Software/scripts/bin/cal_coverage.pl"
rnastat = "/opt/nasapps/production/picard/1.84/CollectRnaSeqMetrics.jar"
rsem = "/opt/nasapps/development/rsem/1.2.22/bin/rsem-calculate-expression"
rsemcounts = "/is2/projects/CCR-SF/active/Software/scripts/R_scripts/rsemcounts.R"
salmon = "/opt/nasapps/development/salmon/0.6.0/bin/salmon"
sambamba = "/opt/nasapps/development/speedseq/0.1.0/bin/sambamba"
star = "/opt/nasapps/development/star/2.5/bin/STAR"
trimmomatic = "/is2/projects/CCR-SF/active/Software/tools/trimmomatic/Trimmomatic-0.30/trimmomatic-0.30.jar"
wkthtmltopdf = "/is2/projects/CCR-SF/active/Software/tools/wkhtmltopdf-amd64"